include/mockturtle/io/aiger_reader.hpp

```diff
@@ -106,19 +160,157 @@ class aiger_reader : public lorina::aiger_reader
   void on_input_name( unsigned index, const std::string& name ) const override
   {
     if ( _names )
+      _names->insert( signals[1 + index], name );
+  }
+
+  void on_output_name( unsigned index, const std::string& name ) const override
+  {
+    std::get<1>( outputs[index] ) = name;
+  }
+
+  void on_and( unsigned index, unsigned left_lit, unsigned right_lit ) const override
+  {
+    (void)index;
+    assert( signals.size() == index );
+
+    auto left = signals[left_lit >> 1];
+    if ( left_lit & 1 )
+    {
+      left = _ntk.create_not( left );
+    }
+
+    auto right = signals[right_lit >> 1];
+    if ( right_lit & 1 )
     {
-      _names->insert( std::make_pair(signals[1 + index], name) );
+      right = _ntk.create_not( right );
     }
+
+    signals.push_back( _ntk.create_and( left, right ) );
   }
 
-  void on_output_name( unsigned index, const std::string& name ) const override
+  void on_output( unsigned index, unsigned lit ) const override
   {
-    if ( _names )
+    (void)index;
+    assert( index == outputs.size() );
+    outputs.push_back( std::make_tuple( lit, "" ) );
+  }
+
+private:
+  Ntk& _ntk;
+
+  mutable std::vector<std::tuple<unsigned,std::string>> outputs;
+  mutable std::vector<signal<Ntk>> signals;
+  mutable NameMap<Ntk>* _names;
+};
+
+/*! \brief Lorina reader callback for Aiger files.
+ *
+ * **Required network functions:**
+ * - `create_pi`
+ * - `create_po`
+ * - `create_ro`
+ * - `create_ri`
+ * - `get_constant`
+ * - `create_not`
+ * - `create_and`
+ *
+   \verbatim embed:rst
+
+   Example
+
+   .. code-block:: c++
+
+      aig_network aig;
+      lorina::read_aiger( "file.aig", aiger_reader( aig ) );
+
+      mig_network mig;
+      lorina::read_aiger( "file.aig", aiger_reader( mig ) );
+   \endverbatim
+ */
+template<>
+class aiger_reader<aig_network> : public lorina::aiger_reader
```

Comment:
This is too restrictive. Instead I suggest to have a trait, e.g., `implements_sequential_interface<Ntk>` that checks whether a certain set of network methods are implemented. Then this would enable this version of the aiger_reader, or with `if constexpr` in the original implementation. Sooner or later, the interface becomes implemented for MIGs, XMGs, ..., and then this reader would work as well for them.